lib/oauth2-client.js

```diff
@@ -0,0 +1,131 @@
+/*!
+ * Copyright (c) 2021 Digital Bazaar, Inc. All rights reserved.
+ */
+
+import database from 'bedrock-mongodb';
+import * as bedrock from 'bedrock';
+import {promisify} from 'util';
+import {httpClient} from '@digitalbazaar/http-client';
+const {util: {BedrockError}} = bedrock;
+
+bedrock.events.on('bedrock-mongodb.ready', async () => {
+  await promisify(database.openCollections)(['access-tokens']);
+
+  // shard based on `siteId`
+  await promisify(database.createIndexes)([{
+    // prevent duplicate events from being stored
+    collection: 'access-tokens',
+    fields: {'token.token': 1},
+    options: {unique: true, background: false}
+  }, {
+    // automatically expire events with an `expires` date field
+    collection: 'access-tokens',
+    fields: {'token.expires': 1},
+    options: {
+      partialFilterExpression: {'token.expires': {$exists: true}},
+      unique: false,
+      background: false,
+      expireAfterSeconds: 0
+    }
+  }]);
+});
+
+export class OAuth2Client {
+  constructor({
+    client_id, client_secret, url, grant_type, scope, httpsAgent
+  } = {}) {
+    this.client_id = client_id;
+    this.client_secret = client_secret;
+    this.url = url;
+    this.grant_type = grant_type;
+    this.scope = scope;
+    this.httpsAgent = httpsAgent;
+  }
+  async getToken() {
+    const {client_id, client_secret, url, grant_type, scope, httpsAgent} = this;
+    let token;
+    token = await checkDatabase({client_id});
+    if(!token || !token.access_token) {
+      token = await getNewAccessToken({
+        client_id, client_secret, url, grant_type, scope, httpsAgent
+      });
+      await insertToken({token});
+    }
+    return token.access_token;
+  }
+}
+
+async function getNewAccessToken({
+  client_id, client_secret, url, grant_type, scope, httpsAgent
+}) {
+  const body = new URLSearchParams();
+  body.set('client_id', client_id);
+  body.set('client_secret', client_secret);
+  body.set('grant_type', grant_type);
+  body.set('scope', scope);
+
+  let response;
+  try {
+    ({data: response} = await httpClient.post(url, {
+      body,
+      agent: httpsAgent
+    }));
+  } catch(e) {
+    throw new BedrockError(
```

Comment:
Can we add an error log entry here as well, in addition to throwing the BedrockError?
Returning an opaque 503 Service Unavailable to whoever is trying to use this endpoint is totally fine,
but we also want to make sure that our error logs contain the reason "This service is unavailable because we couldn't get a critical access_token - and here's the reason that operation failed."

Reply:
Added logger.